Explain why the AM becomes group leader

fxn opened the pull request:
I think it is not obvious for the reader of the code why do application masters become group leader of their applications.

You can kind of infer it if you read a lot, connect the dots, and allow yourself to conclude that this is the intention despite it does not seem motivated by I/O... but if you are told right away, you not only get the picture upfront, but also have official confimation rather than leaving yourself with a conjecture :).

The comment itself comes in part from my own research, and also from feedback from @ferd in the mailing list (thanks again!).

fxn commented:
BTW, the proposed code comment matches the assumptions in the code, but let me say too that I find these assumptions strange.

For example, take this function in `applicaction`:

```erlang
get_application() -> 
    application_controller:get_application(group_leader()).
```

it calls to

```erlang
get_application(Master) ->
    case ets:match(ac_tab, {{application_master, '$1'}, Master}) of
	[[AppName]] -> {ok, AppName};
	_ -> undefined
    end.
```

That matches what the proposed code comment says. Good.

But, there is [API to change the group leader](http://erlang.org/doc/man/erlang.html#group_leader-2). The documentation of `group_leader/2` gives a typical use case, but it does not say to users they shouldn't call this function. As far as the API is concerned, this is something programs can change.

So, I am finding something missing: Either the assumption that the application master of an application is the group leader of its process tree is broken because users may change group leaders at any time, or else the documentation of `group_leader/2` should warn and restrict its public contract somehow.

What do you think?

sirihansen commented:
Thank you, this information is really missing and, as you point out, should also be added to the documentation. We would be very happy for a contribution on this.

From what I see, at least some facts should be added under the [design principles](https://github.com/erlang/otp/blob/c171aa54085853afe4bf7097c9b47a0fa9a72378/system/doc/design_principles/applications.xml#L365) and in the [ref.man. for application](https://github.com/erlang/otp/blob/c171aa54085853afe4bf7097c9b47a0fa9a72378/lib/kernel/doc/src/application.xml#L320). And the [ref.man. for group_leader/2](https://github.com/erlang/otp/blob/c171aa54085853afe4bf7097c9b47a0fa9a72378/erts/doc/src/erlang.xml#L2063) should warn and link to the above two places.

fxn commented:
Awesome, I’ll volunteer a patch!

sirihansen commented:
Great! Were you planning on updating this PR or adding a new one? For us it would be fine to have it all in one place (i.e. in this PR).

fxn commented:
@sirihansen hey! I have gone through those other files too.

I have left the code comment as a micro-decision, but please don't hesitate to ask for its removal if you prefer it. Also, don't hesitate to ask for any rewording or whatever.

sirihansen commented on erts/doc/src/erlang.xml:
```diff
@@ -2068,6 +2068,10 @@ end</pre>
           Typically, this is used when a process started from a
           certain shell is to have another group leader than
           <c>init</c>.</p>
+        <p>The process leader should be rarely changed in
```
Please change "process leader" and "project leader" to "group leader". And maybe we should be a bit more firm? A note? @bjorng , what do you think? Or at least use "must not" instead of "should rarely"?
No details must be written here, but please add link to the design principles (use `seealso` tag, with attribute `marker="doc/design_principles:applications#stopping"`).

fxn commented:
@sirihansen can't believe I wrote "project leader", go figure 😄. Thanks for the review, I have addressed those mistakes and added the see also link.

The rebase deleted the question for @bjorng though, I recover it here from the notification email:

> And maybe we should be a bit more firm? A note? @bjorng , what do you think? Or at least use "must not" instead of "should rarely"?

sirihansen commented:
About to merge this, but just one question: what is the meaning of `[ci skip]` in the commit message?

And, could you please change the first letter of the commit message to upper case? Thanks!

fxn commented:
Fantastic! About `[ci skip]`, some CIs check for this, and if present the suite does not run. Handy when CI is busy and the patch touches no code.

But if you don’t use that I can also amend it, wouldn’t make sense in your history. What do you prefer?

sirihansen commented:
Right - we don't use that, so it would look better without it, I think. Thanks!

fxn commented:
@sirihansen all up and green!
